composer/trainer/dist_strategy.py

```diff
@@ -369,78 +369,141 @@ def sync_hook(*args):
-                                   'the tied weights are in the same FSDP module.'))
-
-                # Redoes weight tying
-                for name_attr, tied_names in tied_mod_names.items():
-                    name, attr = name_attr
-                    src_mod = module.get_submodule(name)
-                    # We need to make sure the source and destination
-                    # modules end up in the same FSDP module otherwise
-                    # with sharding weight tying gets violated
-                    src_mod._fsdp_wrap = False  # type: ignore
-                    src_params = getattr(src_mod, attr)
-                    for tied_name in tied_names:
-                        dest_mod = module.get_submodule(tied_name)
-                        dest_mod._fsdp_wrap = False  # type: ignore
-                        setattr(dest_mod, attr, src_params)
-
-                if hasattr(obj, 'param_init_fn') and isinstance(obj.param_init_fn, Callable):
-                    module.apply(obj.param_init_fn)
-                elif hasattr(module, 'reset_parameters') and isinstance(module.reset_parameters, Callable):
-                    module.reset_parameters()
-                else:
-                    raise ValueError(
-                        f'Object `{obj_name}` does not have a ``param_init_fn`` or a ``reset_parameters`` function. '
-                        'This leaves parameters without initialization. Please add a ``param_init_fn`` or ``reset_parameters`` '
-                        f'to module `{obj_name}`.')
+
+                    # Arbitrarily choose the first module as the source module
+                    first_mod, first_attr = mod_attr_list[0]
+                    source_mod_to_mod_attr[first_mod] = [
+                        (target_mod, first_attr, dest_attr) for target_mod, dest_attr in mod_attr_list[1:]
+                    ]
+
+                # Clean up no longer needed module references for memory safety
+                del tied_pointers
+
+                def _param_init_fn(module: torch.nn.Module) -> None:
+                    # If we do not have any parameters or buffers on meta device managed by this module directly, we do not need to call the parameter init function.
+                    # It is assumed that whatever process moved the parameters off of meta device initialized them.
+                    # We expect this to occur if we have tied weights, as the second module will already have the weights initialized.
+                    is_meta = any(param.is_meta for param in module.parameters(recurse=False)) or any(
+                        buffer.is_meta for buffer in module.buffers(recurse=False))
+                    if not is_meta:
+                        return
+
+                    # Move all parameters and buffers to the current device
+                    module.to_empty(device=f'cuda:{torch.cuda.current_device()}', recurse=False)
```

Comment:
I don't think so. Here is my logic.

I call `to_empty` with `recurse=False` so that it does _not_ initialize submodules. Above this I check if the module still has params/buffers on `meta` (again with `recurse=False`) before calling `to_empty`, so `to_empty` should not get called on anything that has already been moved off of `meta` device. Let me know what you think.

Reply:
I think this makes sense. Thanks for clarifying!